run.py

```diff
@@ -1,345 +1,401 @@
+        log.debug("%s exists.", FREESURFER_LICENSE)
+    install_freesurfer_license(gtk_context, FREESURFER_LICENSE)
+
+    command = generate_command(
+        config, gtk_context.work_dir, output_analysis_id_dir, log, errors, warnings
+    )
+
+    # This is used as part of the name of output files
+    command_name = make_file_name_safe(command[0])
+
+    # Download BIDS Formatted data
+    if len(errors) == 0:
+
+        # Create HTML file that shows BIDS "Tree" like output
+        tree = True
+        tree_title = f"{command_name} BIDS Tree"
+
+        error_code = download_bids_for_runlevel(
+            gtk_context,
+            hierarchy,
+            tree=tree,
+            tree_title=tree_title,
+            src_data=DOWNLOAD_SOURCE,
+            folders=DOWNLOAD_MODALITIES,
+            dry_run=dry_run,
+            do_validate_bids=config.get("gear-run-bids-validation"),
+        )
+        if error_code > 0 and not config.get("gear-ignore-bids-errors"):
+            errors.append(f"BIDS Error(s) detected.  Did not run {CONTAINER}")
 
-        if len(context.gear_dict['errors']) > 0:
-            ok_to_run = False
-            result = sp.CompletedProcess
-            result.returncode = 1
-            log.info('Command was NOT run because of previous errors.')
+    else:
+        log.info("Did not download BIDS because of previous errors")
+        print(errors)
 
-        elif context.config['gear-dry-run']:
-            ok_to_run = False
-            result = sp.CompletedProcess
-            result.returncode = 0
-            e = 'gear-dry-run is set: Command was NOT run.'
-            log.warning(e)
-            context.gear_dict['warnings'].append(e)
-            utils.dry_run.pretend_it_ran(context)
+    # Don't run if there were errors or if this is a dry run
+    return_code = 0
 
-        if ok_to_run:
-            # Run the actual command this gear was created for
-            result = sp.run(context.gear_dict['command'], 
-                        env = context.gear_dict['environ'])
-            log.debug(repr(result))
+    if len(errors) > 0:
+        return_code = 1
+        log.info("Command was NOT run because of previous errors.")
 
-        log.info('Return code: ' + str(result.returncode))
+    try:
 
-        if result.returncode == 0:
-            log.info('Command successfully executed!')
+        if dry_run:
+            e = "gear-dry-run is set: Command was NOT run."
+            log.warning(e)
+            warnings.append(e)
+            pretend_it_ran(gtk_context)
 
         else:
-            log.info('Command failed.')
+            # Create output directory
+            log.info("Creating output directory %s", output_analysis_id_dir)
+            Path(output_analysis_id_dir).mkdir()
 
-    except Exception as e:
-        context.gear_dict['errors'].append(e)
-        log.critical(e)
-        log.exception('Unable to execute command.')
+            # This is what it is all about
```

Comment:
For fMRIPrep, that might be better.  But for general bids apps, I want to exit after uploading whatever might have been saved (like the bids tree output) in case it is helpful to figure out what the gear failed.